File: src/0238-Product-of-Array-Except-Self/product_of_array_except_selfs.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<int> productExceptSelf(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<int> left_products (n, 1);
        std::vector<int> right_products (n, 1);
        std::vector<int> ans(n, 1);
        
        for (int i = 1; i < n; i++) {
            left_products [i] = left_products [i - 1] * nums[i - 1];
        }
        
        for (int i = n - 2; i >= 0; i--) {
            right_products [i] = right_products [i + 1] * nums[i + 1];
        }
        
        for (int i = 0; i < n; i++) {
            ans[i] = left_products [i] * right_products [i];
        }
        
        return ans;
    }
};

int main() {
    std::vector<int> nums = {1,2,3,4};
    Solution solution;

    std::vector<int> output =  solution.productExceptSelf(nums);

    for (auto it : output) {
        std::cout << it << " ";
    }

    return 0;
}

File: src/0238-Product-of-Array-Except-Self/product_of_array_except_selfs.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left_products = [1] * n
        right_products = [1] * n
        ans = [1] * n
        
        for i in range(1, n):
            left_products[i] = left_products[i - 1] * nums[i - 1]
        
        for i in range(n - 2, -1, -1):
            right_products[i] = right_products[i + 1] * nums[i + 1]
        
        for i in range(n):
            ans[i] = left_products[i] * right_products[i]
        
        return ans

if __name__ == "__main__":
    nums: List[int] = [1,2,3,4]
    solution = Solution()

    print(solution.productExceptSelf(nums))